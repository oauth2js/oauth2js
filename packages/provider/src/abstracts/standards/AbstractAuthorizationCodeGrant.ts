import { Mixed } from '@prunus/mixin'

import { AccessDenied } from '../../errors/AccessDenied'
import { InvalidClient } from '../../errors/InvalidClient'
import { InvalidGrant } from '../../errors/InvalidGrant'
import { InvalidRequest } from '../../errors/InvalidRequest'
import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { CODE_REGEXP } from '../../helpers/regexps'
import { Maybe, PromiseToo } from '../../helpers/types'
import { IAuthorizationCode } from '../../interfaces/IAuthorizationCode'
import { IClient } from '../../interfaces/IClient'
import { RequestLocation } from '../../interfaces/IRequest'
import { IScope } from '../../interfaces/IScope'
import { ITicket } from '../../interfaces/ITicket'
import { IToken } from '../../interfaces/IToken'
import { IUser } from '../../interfaces/IUser'
import { Context } from '../../models/Context'
import {
  ClientCredentials,
  UseClientByCredentials,
} from '../../traits/UseClientByCredentials'
import { UseClientById } from '../../traits/UseClientById'
import { UseClientId } from '../../traits/UseClientId'
import { UseClientPolymorph } from '../../traits/UseClientPolymorph'
import { UseClientWithRedirectUris } from '../../traits/UseClientWithRedirectUris'
import { UseCodeGenerator } from '../../traits/UseCodeGenerator'
import { UseCodeLifetimeGenerator } from '../../traits/UseCodeLifetimeGenerator'
import { UseRedirectUri } from '../../traits/UseRedirectUri'
import { UseScopes } from '../../traits/UseScopes'
import { UseState } from '../../traits/UseState'
import { UseUserByAuthorizationCode } from '../../traits/UseUserByAuthorizationCode'
import { AbstractExchanger } from '../AbstractExchanger'
import { AbstractFlow } from '../AbstractFlow'
import { OAuth2Error } from '../OAuth2Error'

export abstract class AbstractAuthorizationCodeFlow extends Mixed(
  AbstractFlow,
  UseClientWithRedirectUris(
    UseClientPolymorph(UseClientByCredentials, UseClientById),
  ),
  UseScopes,
  UseRedirectUri,
  UseCodeGenerator,
  UseCodeLifetimeGenerator,
  UseState,
) {
  protected override readonly type = 'code'
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Query
  protected readonly clientCredentialsLocation: RequestLocation =
    RequestLocation.Query
  protected readonly stateLocation: RequestLocation = RequestLocation.Query
  protected readonly allowEmptyState: boolean = false
  protected readonly scopeLocation = RequestLocation.Query
  protected readonly redirectUriLocation = RequestLocation.Query
  protected readonly codeLifetime = 5 * 60 * 1000

  protected override async getRedirectUri(context: Context): Promise<string> {
    const maybeRedirectUri = await super.getRedirectUri(context)

    if (!maybeRedirectUri) return this.getClientDefaultRedirectUri(context)
    return maybeRedirectUri
  }

  protected override assertRedirectUri(
    redirectUri: Maybe<string>,
    context: Context,
  ): asserts redirectUri is string {
    super.assertRedirectUri(redirectUri, context)

    if (!this.getClientRedirectUrisSync(context).includes(redirectUri))
      throw new InvalidRequest(
        'The redirect uri must be whitelisted by the client.',
        context,
      )
  }

  protected async createTicket(context: Context): Promise<IAuthorizationCode> {
    const client = await this.getOrLoadClient(context)
    const redirectUri = await this.getOrLoadRedirectUri(context)
    const scopes = await this.getOrLoadScopesObjects(context)
    const [code, codeExpiresAt, state] = await Promise.all([
      this.generateCode(context),
      this.getCodeExpiresAt(context),
      this.getState(context),
    ])

    return { client, code, codeExpiresAt, scopes, redirectUri, state }
  }

  protected async release(ticket: IAuthorizationCode, context: Context) {
    const authorizationCode = await this.saveAuthorizationCode(ticket, context)

    await this.sendAuthorizationCode(authorizationCode, context)
  }

  protected buildURI(uri: string, context: Context): PromiseToo<URL>
  protected buildURI(uri: string): PromiseToo<URL> {
    const data = new URL(uri)

    data.search = ''
    return data
  }

  protected redirect(uri: URL, context: Context) {
    context.response.setHeader('Location', uri.toString())
    context.response.status(302)
    context.response.send()
  }

  protected async authorize(
    ticket: IAuthorizationCode,
    user: IUser,
    context: Context,
  ): Promise<void> {
    await this.authorizeAuthorizationCode(ticket, user, context)

    const uri = await this.buildURI(ticket.redirectUri, context)

    uri.searchParams.set('code', ticket.code)
    if (ticket.state) uri.searchParams.set('state', ticket.state)
    this.redirect(uri, context)
  }

  protected async deny(
    ticket: IAuthorizationCode,
    context: Context,
  ): Promise<void> {
    await this.denyAuthorizationCode(ticket, context)
    throw new AccessDenied('Access is denied', context)
  }

  protected async releaseError(error: OAuth2Error, context: Context) {
    if (error instanceof InvalidClient)
      await this.sendErrorWithInvalidClient(error, context)

    try {
      await this.getOrLoadClient(context)
    } catch {
      return await this.sendErrorWithInvalidClient(error, context)
    }

    try {
      const redirectUri = await this.getOrLoadRedirectUri(context)
      const uri = await this.buildURI(redirectUri, context)

      uri.searchParams.set('error', error.error)
      uri.searchParams.set('error_description', error.description)
      this.redirect(uri, context)
    } catch {
      await this.sendErrorWithInvalidRedirectUri(error, context)
    }
  }

  protected async authorizeError(
    error: OAuth2Error,
    ticket: IAuthorizationCode,
    user: IUser,
    context: Context,
  ) {
    const uri = await this.buildURI(ticket.redirectUri, context)

    uri.searchParams.set('error', error.error)
    uri.searchParams.set('error_description', error.description)
    this.redirect(uri, context)
  }

  protected async denyError(
    error: OAuth2Error,
    ticket: IAuthorizationCode,
    context: Context,
  ) {
    const uri = await this.buildURI(ticket.redirectUri, context)

    uri.searchParams.set('error', error.error)
    uri.searchParams.set('error_description', error.description)
    this.redirect(uri, context)
  }

  protected abstract saveAuthorizationCode(
    authorizationCode: IAuthorizationCode,
    context: Context,
  ): PromiseToo<IAuthorizationCode>

  protected abstract authorizeAuthorizationCode(
    authorizationCode: IAuthorizationCode,
    user: IUser,
    context: Context,
  ): PromiseToo<void>

  protected abstract denyAuthorizationCode(
    authorizationCode: IAuthorizationCode,
    context: Context,
  ): PromiseToo<void>

  protected abstract sendErrorWithInvalidClient(
    error: OAuth2Error,
    context: Context,
  ): PromiseToo<void>

  protected abstract sendErrorWithInvalidRedirectUri(
    error: OAuth2Error,
    context: Context,
  ): PromiseToo<void>

  protected abstract sendAuthorizationCode(
    authorizationCode: IAuthorizationCode,
    context: Context,
  ): PromiseToo<void>

  protected abstract override selectUseClientProvider(
    context: Context,
  ): typeof UseClientByCredentials | typeof UseClientById

  protected abstract override getClientById(
    clientId: string,
    context: Context,
  ): PromiseToo<Maybe<IClient>>

  protected abstract override getClientByCredentials(
    credentials: ClientCredentials,
    context: Context,
  ): PromiseToo<Maybe<IClient>>

  protected abstract override getScopesObjects(
    scopes: string[],
    context: Context,
  ): PromiseToo<IScope[]>
}

const ClassMix = Mixed(
  AbstractExchanger,
  UseRedirectUri,
  UseClientId,
  UseUserByAuthorizationCode,
)

export abstract class TT extends ClassMix {}

export abstract class AbstractAuthorizationCodeGrant extends Mixed(
  AbstractExchanger,
  UseRedirectUri,
  UseClientId,
  UseUserByAuthorizationCode,
) {
  protected readonly refreshTokenLifetime = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly accessTokenLifetime = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body
  protected readonly redirectUriLocation: RequestLocation = RequestLocation.Body

  protected abstract override readonly flow: AbstractAuthorizationCodeFlow

  protected revokeTicket(
    ticket: IAuthorizationCode,
    context: Context,
  ): PromiseToo<boolean> {
    return this.revokeAuthorizationCode(ticket, context)
  }

  protected getAuthorizationCode(
    context: Context,
  ): PromiseToo<IAuthorizationCode> {
    return this.getOrLoadTicket(context) as PromiseToo<IAuthorizationCode>
  }

  protected async exchangeTicketByToken(
    authorizationCode: IAuthorizationCode,
    context: Context,
  ): Promise<IToken> {
    if ((await this.getClientId(context)) !== authorizationCode.client.id)
      throw new InvalidClient('Client identifier does not match', context)

    if ((await this.getRedirectUri(context)) !== authorizationCode.redirectUri)
      throw new InvalidGrant('Redirect uri does not match', context)

    const client = authorizationCode.client

    const user = await this.getOrLoadUser(context)

    const scopes = authorizationCode.scopes

    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      user,
      client,
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
      scopes,
    }
  }

  protected getCode(context: Context): PromiseToo<string> {
    if (!context.request.body.code)
      throw new InvalidRequest('The code is missing', context)

    if (typeof context.request.body.code !== 'string')
      throw new InvalidRequest(
        `The code must be of type string but is defined as ${typeof context
          .request.body.code}`,
        context,
      )

    // TODO: Invalid request or grant?
    if (!CODE_REGEXP.test(context.request.body.code))
      throw new InvalidRequest('The code is invalid', context)

    return context.request.body.code
  }

  protected async getTicket(context: Context): Promise<Maybe<ITicket>> {
    const code = await this.getCode(context)

    return this.getAuthorizationCodeByCode(code, context)
  }

  protected abstract override saveExchangeToken(
    token: IToken,
    ticket: ITicket,
    context: Context,
  ): PromiseToo<IToken>

  protected abstract getAuthorizationCodeByCode(
    code: string,
    context: Context,
  ): PromiseToo<Maybe<IAuthorizationCode>>

  protected abstract revokeAuthorizationCode(
    authorizationCode: IAuthorizationCode,
    context: Context,
  ): PromiseToo<boolean>
}
