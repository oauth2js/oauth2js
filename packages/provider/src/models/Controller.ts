import { is } from 'type-is'

import { AbstractExchanger } from '../abstracts/AbstractExchanger'
import { AbstractFlow } from '../abstracts/AbstractFlow'
import { AbstractGrant } from '../abstracts/AbstractGrant'
import { AbstractTokenType } from '../abstracts/AbstractTokenType'
import { isOAuth2Error } from '../abstracts/OAuth2Error'
import { InvalidRequest } from '../errors/InvalidRequest'
import { ServerError } from '../errors/ServerError'
import { UnsupportedGrantType } from '../errors/UnsupportedGrantType'
import { GRANT_TYPE_REGEXP, RESPONSE_TYPE_REGEXP } from '../helpers/regexps'
import {
  ArrayInfer,
  ArrayReadOnlyToo,
  ConstructorOf,
  Maybe,
} from '../helpers/types'
import { IMPLEMENTATION_TYPE } from '../interfaces/IControllerImplementation'
import { IRequest } from '../interfaces/IRequest'
import { IResponse } from '../interfaces/IResponse'
import { ITicket } from '../interfaces/ITicket'
import { IUser } from '../interfaces/IUser'
import { TokenBearerType } from '../standards/TokenBearerType'
import { Context } from './Context'

export type Index = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  -1,
]

export type WalkAnd<
  T extends ArrayReadOnlyToo<ConstructorOf<unknown>>,
  K extends string | symbol,
  I extends number = 0,
> = I extends -1
  ? never
  : T[I] extends undefined
  ? never
  : K extends keyof InstanceType<T[I]>
  ? T[Index[I]] extends undefined
    ? InstanceType<T[I]>[K]
    : InstanceType<T[I]>[K] | WalkAnd<T, K, Index[I]>
  : never

export type ImplementationOf<
  G extends
    | ArrayReadOnlyToo<ConstructorOf<AbstractGrant>>
    | ConstructorOf<AbstractGrant>,
> = G extends ArrayReadOnlyToo<ConstructorOf<AbstractGrant>>
  ? Partial<WalkAnd<G, IMPLEMENTATION_TYPE>>
  : G extends ConstructorOf<AbstractGrant>
  ? Partial<WalkAnd<[G], IMPLEMENTATION_TYPE>>
  : never

export class Controller<
  G extends ArrayReadOnlyToo<ConstructorOf<AbstractGrant>> = [],
> {
  private _grants: AbstractGrant[] = []
  private _flows: AbstractFlow[] = []
  private _tokenTypes: AbstractTokenType[] = [new TokenBearerType()]

  public use<T extends ConstructorOf<AbstractGrant>>(
    Grant: T,
  ): Controller<[...G, T]> {
    const grant = new Grant()

    this._grants.push(grant)

    grant['whenBeingAttached'](this as Controller)

    if (grant instanceof AbstractExchanger && 'flow' in grant) {
      const flow = grant['flow']

      this._flows.push(flow)

      flow['whenBeingAttached'](this as Controller)
    }

    return this as unknown as Controller<[...G, T]>
  }

  public implement(implementations: ImplementationOf<G>): this
  public implement<T extends ArrayReadOnlyToo<ArrayInfer<G>>>(
    implementations: ImplementationOf<T>,
    targets: T,
  ): this
  public implement(
    implementations: ImplementationOf<G>,
    targets?: Array<ConstructorOf<AbstractGrant>>,
  ): this {
    if (targets)
      this._grants
        .filter(grant =>
          targets.includes(grant.constructor as ConstructorOf<AbstractGrant>),
        )
        .forEach(instance => Object.assign(instance, implementations))
    else
      this._grants.forEach(instance => Object.assign(instance, implementations))

    return this
  }

  public async token(request: IRequest, response: IResponse): Promise<void> {
    const context = new Context(request, response)

    try {
      if (
        !(
          is(
            context.request.headers['content-type'] as string,
            'application/x-www-form-urlencoded',
          ) ||
          is(
            context.request.headers['Content-Type'] as string,
            'application/x-www-form-urlencoded',
          )
        )
      )
        throw new InvalidRequest(
          "Requests for an access token must be in the form 'application/x-www-form-urlencoded'",
          context,
        )

      const grantType = request.body.grant_type

      if (!grantType)
        throw new InvalidRequest('The grant type is missing', context)

      if (typeof grantType !== 'string')
        throw new InvalidRequest(
          `The grant type must be of type string but is set to ${typeof grantType}`,
          context,
        )

      // TODO: Invalid request or grant?
      if (!GRANT_TYPE_REGEXP.test(grantType))
        throw new InvalidRequest('The grant type is invalid', context)

      const grant = this._grants.find(grant => grant['type'] === grantType)

      if (!grant)
        throw new UnsupportedGrantType(
          'The grant type is not currently supported by the server.',
          context,
        )

      const token = await AbstractGrant.token.call(grant, context)

      const tokenType = this._tokenTypes.find(
        tokenType => tokenType.type === 'bearer',
      )

      if (!tokenType)
        throw new ServerError('No response type implemented on server', context)

      context.response.setHeader('Content-Type', 'application/json')
      context.response.status(200)
      context.response.send(JSON.stringify(tokenType.toJSON(token)))
    } catch (error: unknown) {
      if (isOAuth2Error(error)) {
        context.response.setHeader('Content-Type', 'application/json')
        context.response.status(error.status)
        context.response.send(
          JSON.stringify({
            error: error.error,
            error_description: error.description,
          }),
        )
      } else throw error
    }
  }

  public async release(request: IRequest, response: IResponse): Promise<void> {
    const context = new Context(request, response)
    const responseType =
      request.body.response_type ?? request.query.response_type

    if (!responseType)
      throw new InvalidRequest('The response type is missing', context)

    if (typeof responseType !== 'string')
      throw new InvalidRequest(
        `The response type must be of type string but is set to ${typeof responseType}`,
        context,
      )

    // TODO: Invalid request or grant?
    if (!RESPONSE_TYPE_REGEXP.test(responseType))
      throw new InvalidRequest('The response type is invalid', context)

    const flow = this._flows.find(flow => flow['type'] === responseType)

    if (!flow)
      throw new UnsupportedGrantType(
        'The response type is not currently supported by the server.',
        context,
      )

    await AbstractFlow.release.call(flow, context).catch(async error => {
      if (!isOAuth2Error(error)) throw error
      await AbstractFlow.releaseError.call(flow, error, context)
    })
  }

  public async authorize(
    exchanger: ConstructorOf<AbstractExchanger>,
    ticket: ITicket,
    user: IUser,
    request: IRequest,
    response: IResponse,
  ) {
    const context = new Context(request, response)

    const grant = this._grants.find(
      grant => grant instanceof exchanger,
    ) as Maybe<AbstractExchanger>

    if (!grant)
      throw new ServerError(
        `No instance of ${exchanger.name} grant found`,
        context,
      )

    const flow = grant['flow'] as Maybe<AbstractFlow>

    if (!flow)
      throw new ServerError(
        `The ${exchanger.name} grant has no ticket flow.`,
        context,
      )

    await AbstractFlow.authorize
      .call(flow, ticket, user, context)
      .catch(async error => {
        if (!isOAuth2Error(error)) throw error
        await AbstractFlow.authorizeError.call(
          flow,
          error,
          ticket,
          user,
          context,
        )
      })
  }

  public async deny(
    exchanger: ConstructorOf<AbstractExchanger>,
    ticket: ITicket,
    request: IRequest,
    response: IResponse,
  ) {
    const context = new Context(request, response)
    const grant = this._grants.find(
      grant => grant instanceof exchanger,
    ) as Maybe<AbstractExchanger>

    if (!grant)
      throw new ServerError(
        `No instance of ${exchanger.name} grant found`,
        context,
      )

    const flow = grant['flow'] as Maybe<AbstractFlow>

    if (!flow)
      throw new ServerError(
        `The ${exchanger.name} grant has no ticket flow.`,
        context,
      )

    await AbstractFlow.deny.call(flow, ticket, context).catch(async error => {
      if (!isOAuth2Error(error)) throw error
      await AbstractFlow.denyError.call(flow, error, ticket, context)
    })
  }
}
