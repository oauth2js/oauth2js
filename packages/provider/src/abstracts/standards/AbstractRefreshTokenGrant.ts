import { Mixed } from '@prunus/mixin'

import { InvalidClient } from '../../errors/InvalidClient'
import { InvalidGrant } from '../../errors/InvalidGrant'
import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { Maybe, PromiseToo } from '../../helpers/types'
import { RequestLocation } from '../../interfaces/IRequest'
import { IToken } from '../../interfaces/IToken'
import { Context } from '../../models/Context'
import { UseClientId } from '../../traits/UseClientId'
import { UseTokenByRefreshToken } from '../../traits/UseTokenByRefreshToken'
import { AbstractGrant } from '../AbstractGrant'

export abstract class AbstractRefreshTokenGrant extends Mixed(
  AbstractGrant,
  UseTokenByRefreshToken,
  UseClientId,
) {
  protected accessTokenLifetime: number = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected refreshTokenLifetime: number = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body

  protected readonly refreshTokenLocation: RequestLocation =
    RequestLocation.Body

  protected async createToken(context: Context): Promise<IToken> {
    const token = await this.getOrLoadTokenByRefreshToken(context)

    if ((await this.getClientId(context)) !== token.client.id)
      throw new InvalidClient('Client identifier does not match', context)

    const { client, user, scopes } = token
    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      client,
      user,
      scopes,
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    }
  }

  protected async saveToken(token: IToken, context: Context): Promise<IToken> {
    return await this.saveTokenByRefreshedToken(
      token,
      await this._revokeToken(context),
      context,
    )
  }

  private async _revokeToken(context: Context) {
    const tokenToRevoke = await this.getOrLoadTokenByRefreshToken(context)

    if (!(await this.revokeToken(tokenToRevoke, context)))
      throw new InvalidGrant('The token is invalid', context)

    return tokenToRevoke
  }

  protected abstract override getTokenByRefreshToken(
    refreshToken: string,
    context: Context,
  ): PromiseToo<Maybe<IToken>>

  protected abstract saveTokenByRefreshedToken(
    token: IToken,
    refreshedToken: IToken,
    context: Context,
  ): PromiseToo<IToken>

  protected abstract revokeToken(
    token: IToken,
    context: Context,
  ): PromiseToo<boolean>
}
