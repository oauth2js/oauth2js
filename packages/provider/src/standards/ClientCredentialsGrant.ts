import { AbstractClientCredentialsGrant } from '../abstracts/standards/AbstractClientCredentialsGrant'
import { ServerError } from '../errors/ServerError'
import { PromiseToo, Maybe } from '../helpers/types'
import { IClient } from '../interfaces/IClient'
import { IControllerImplementation } from '../interfaces/IControllerImplementation'
import { IScope } from '../interfaces/IScope'
import { IToken } from '../interfaces/IToken'
import { IUser } from '../interfaces/IUser'
import { Context } from '../models/Context'
import {
  UseClientByCredentials,
  ClientCredentials,
} from '../traits/UseClientByCredentials'
import { UseClientById } from '../traits/UseClientById'

export class ClientCredentialsGrant extends AbstractClientCredentialsGrant {
  protected selectUseClientProvider(
    context: Context,
  ): typeof UseClientByCredentials | typeof UseClientById {
    if (
      this.getClientByCredentials !==
      ClientCredentialsGrant.prototype.getClientByCredentials
    )
      if (this.getClientById !== ClientCredentialsGrant.prototype.getClientById)
        throw new ServerError(
          'Unable to resolve which client provider to use, conflict when `getClientByCredentials` and `getClientById` are set without a custom `selectUseClientProvider`.',
          context,
        )
      else return UseClientByCredentials

    if (this.getClientById !== ClientCredentialsGrant.prototype.getClientById)
      return UseClientById

    throw new ServerError(
      'Unable to resolve which client provider to use, conflict when neither `getClientByCredentials` nor `getClientById` are defined.',
      context,
    )
  }

  protected getClientByCredentials(
    credentials: ClientCredentials,
    context: Context,
  ): PromiseToo<Maybe<IClient>> {
    throw new ServerError(
      'The `getClientByCredentials` definition is missing from the implementation.',
      context,
    )
  }

  protected getClientById(
    clientId: string,
    context: Context,
  ): PromiseToo<Maybe<IClient>> {
    throw new ServerError(
      'The `getClientById` definition is missing from the implementation.',
      context,
    )
  }

  protected getScopesObjects(
    scopes: string[],
    context: Context,
  ): PromiseToo<IScope[]> {
    throw new ServerError(
      'The `getScopesObjects` definition is missing from the implementation.',
      context,
    )
  }

  protected getUserByClient(
    client: IClient,
    context: Context,
  ): PromiseToo<Maybe<IUser>> {
    throw new ServerError(
      'The `getUserByClient` definition is missing from the implementation.',
      context,
    )
  }

  protected saveToken(token: IToken, context: Context): PromiseToo<IToken> {
    throw new ServerError(
      'The `saveToken` definition is missing from the implementation.',
      context,
    )
  }
}

export interface ClientCredentialsGrant
  extends IControllerImplementation<{
    generateRefreshTokenForClientCredentials: boolean
    getClientByCredentials(
      credentials: ClientCredentials,
      context: Context,
    ): PromiseToo<Maybe<IClient>>
    getClientById(
      clientId: string,
      context: Context,
    ): PromiseToo<Maybe<IClient>>
    getScopesObjects(scopes: string[], context: Context): PromiseToo<IScope[]>
    getUserByClient(client: IClient, context: Context): PromiseToo<Maybe<IUser>>
    saveToken(token: IToken, context: Context): PromiseToo<IToken>
  }> {}
