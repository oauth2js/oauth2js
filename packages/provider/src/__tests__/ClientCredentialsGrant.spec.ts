import { MockController } from '../../.jest/tools/Controller'
import { Database } from '../../.jest/tools/Database'
import {
  DatabaseClient,
  DatabaseScope,
  DatabaseUser,
} from '../../.jest/tools/DatabaseTypes'
import { ClientCredentialsGrant } from '../standards/ClientCredentialsGrant'

describe('Client Credentials Grant', () => {
  it('generate token', async () => {
    const db = Database()

    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
      client: db.clients.insertOne({
        id: 'client-test-id',
        secret: 'client-test-secret',
        redirectUris: ['http://localhost:8080'],
      }),
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller.use(ClientCredentialsGrant).implement({
        getClientByCredentials: credentials =>
          db.clients.findOne<DatabaseClient>(
            client =>
              client.id === credentials.id &&
              client.secret === credentials.secret,
          ),
        getScopesObjects: scopes =>
          db.scopes.find<DatabaseScope>(scope => scopes.includes(scope.alias)),
        getUserByClient: client =>
          db.users.findOne<DatabaseUser>(user => user.client === client),
        saveToken: token => db.tokens.insertOne(token),
      }),
    )

    const headers = {
      Authorization: `Basic ${Buffer.from(
        'client-test-id:client-test-secret',
      ).toString('base64')}`,
    }
    const body = {
      grant_type: 'client_credentials',
      scope: 'read+write',
    }
    const response = await controller.token({ headers, body })

    expect(response.status()).toBe(200)
    expect(response.headers).toEqual({
      'Content-Type': 'application/json',
    })
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
    })
    expect(db.tokens.findOne()).toMatchSnapshot({
      accessToken: expect.any(String),
      accessTokenExpiresAt: expect.any(Date),
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        client: {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      },
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    })
  })

  it('generate token with refresh token', async () => {
    const db = Database()

    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
      client: db.clients.insertOne({
        id: 'client-test-id',
        secret: 'client-test-secret',
        redirectUris: ['http://localhost:8080'],
      }),
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller => {
      controller.use(ClientCredentialsGrant).implement({
        getClientByCredentials: credentials =>
          db.clients.findOne<DatabaseClient>(
            client =>
              client.id === credentials.id &&
              client.secret === credentials.secret,
          ),
        getScopesObjects: scopes =>
          db.scopes.find<DatabaseScope>(scope => scopes.includes(scope.alias)),
        getUserByClient: client =>
          db.users.findOne<DatabaseUser>(user => user.client === client),
        saveToken: token => db.tokens.insertOne(token),
        generateRefreshTokenForClientCredentials: true,
      })
    })

    const headers = {
      Authorization: `Basic ${Buffer.from(
        'client-test-id:client-test-secret',
      ).toString('base64')}`,
    }
    const body = {
      grant_type: 'client_credentials',
      scope: 'read+write',
    }
    const response = await controller.token({ headers, body })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
      refresh_token: expect.any(String),
    })
    expect(db.tokens.findOne()).toMatchSnapshot({
      accessToken: expect.any(String),
      accessTokenExpiresAt: expect.any(Date),
      refreshToken: expect.any(String),
      refreshTokenExpiresAt: expect.any(Date),
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        client: {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      },
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    })
  })
})
