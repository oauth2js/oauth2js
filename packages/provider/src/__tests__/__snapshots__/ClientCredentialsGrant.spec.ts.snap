// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Credentials Grant generate token 1`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": null,
  "token_type": "Bearer",
}
`;

exports[`Client Credentials Grant generate token 2`] = `
Object {
  "accessToken": Any<String>,
  "accessTokenExpiresAt": Any<Date>,
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "createdAt": Any<Date>,
  "deletedAt": null,
  "refreshToken": null,
  "refreshTokenExpiresAt": null,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "user": Object {
    "client": Object {
      "createdAt": Any<Date>,
      "deletedAt": null,
      "id": "client-test-id",
      "redirectUris": Array [
        "http://localhost:8080",
      ],
      "secret": "client-test-secret",
      "updatedAt": Any<Date>,
    },
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;

exports[`Client Credentials Grant generate token with refresh token 1`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Client Credentials Grant generate token with refresh token 2`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": Any<String>,
  "token_type": "Bearer",
}
`;

exports[`Client Credentials Grant generate token with refresh token 3`] = `
Object {
  "accessToken": Any<String>,
  "accessTokenExpiresAt": Any<Date>,
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "createdAt": Any<Date>,
  "deletedAt": null,
  "refreshToken": Any<String>,
  "refreshTokenExpiresAt": Any<Date>,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "user": Object {
    "client": Object {
      "createdAt": Any<Date>,
      "deletedAt": null,
      "id": "client-test-id",
      "redirectUris": Array [
        "http://localhost:8080",
      ],
      "secret": "client-test-secret",
      "updatedAt": Any<Date>,
    },
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;
